controllers/modelmesh.go

```diff
@@ -2,98 +2,114 @@ package controllers
+			for j, e := range deployment.Spec.Template.Spec.Containers[i].Env {
 				if e.Name == envVarName {
-					envVar = &container.Env[i]
+					envVar = &deployment.Spec.Template.Spec.Containers[i].Env[j]
 					break
 				}
 			}
 
 			// If the env var doesn't exist, add it (if we are not removing)
-			if envVar == nil {
-				if !remove {
-					container.Env = append(container.Env, corev1.EnvVar{
-						Name:  envVarName,
-						Value: url,
-					})
-				}
-			} else {
-				// Split the existing value into parts
-				parts := strings.Split(envVar.Value, " ")
-
-				// Check if the payload processor already exists in the list
-				exists := false
-				for _, part := range parts {
-					if part == url {
-						exists = true
-						break
+			if envVar == nil && !remove {
+				deployment.Spec.Template.Spec.Containers[i].Env = append(deployment.Spec.Template.Spec.Containers[i].Env, corev1.EnvVar{
+					Name:  envVarName,
+					Value: url,
+				})
+			} else if envVar != nil {
+				// If the env var exists and already contains the value, don't do anything
+				existingValues := strings.Split(envVar.Value, " ")
+				for _, v := range existingValues {
+					if v == url {
+						continue
 					}
 				}
 
-				// If we are removing, and the payload processor exists, remove it
-				if remove && exists {
-					newParts := []string{}
-					for _, part := range parts {
-						if part != url {
-							newParts = append(newParts, part)
-						}
-					}
-					envVar.Value = strings.Join(newParts, " ")
-				}
+				// Modify the existing env var based on the remove flag and current value
+				envVar.Value = updateEnvVarValue(envVar.Value, url, remove)
+			}
 
-				// If we are not removing, and the URL doesn't exist, add it
-				if !remove && !exists {
-					envVar.Value += " " + url
-				}
+			// Update the Deployment
+			if err := r.Update(ctx, &deployment); err != nil {
+				log.FromContext(ctx).Error(err, "Could not update Deployment", "Deployment", deployment.Name)
+				return false, err
 			}
+			log.FromContext(ctx).Info("Updating Deployment " + deployment.Name + ", container spec " + deployment.Spec.Template.Spec.Containers[i].Name + ", env var " + envVarName + " to " + url)
+		}
+	}
+
+	return true, nil
+}
+
+func (r *TrustyAIServiceReconciler) patchEnvVarsByLabelForDeployments(ctx context.Context, namespace string, labelKey string, labelValue string, envVarName string, crName string, remove bool) (bool, error) {
+	// Get all Deployments for the label
+	deployments, err := r.GetDeploymentsByLabel(ctx, namespace, labelKey, labelValue)
+	if err != nil {
+		log.FromContext(ctx).Error(err, "Could not get Deployments by label.")
+		return false, err
+	}
+
+	// Build the payload processor endpoint
+	url := "http://" + crName + "." + namespace + ".svc.cluster.local/consumer/kserve/v2"
+
+	// Patch environment variables for the Deployments
+	if shouldContinue, err := r.patchEnvVarsForDeployments(ctx, deployments, envVarName, url, remove); err != nil {
```

Comment:
Minor note, I think this code is good but the `remove` codepath is a bit hard to understand: the logic is hidden inside `patchEnvVarsForDeployments` so for example the `url` is built even while removing.
I think it is fine for now, we can probably improve it in the future but I would like to minimise the changes at this stage :)

Reply:
https://github.com/trustyai-explainability/trustyai-service-operator/issues/56